### PR TITLE
Fix several issues

### PR DESCRIPTION
**Pull request by gonzalocasas**
Solves https://github.com/compas-dev/compas_eve/issues/11

This PR fixes several things: first of all, the base case of the `Message` class was removed and re-based to `object` because `UserDict` is an old-style class and that creates a ton of problems.

### What type of change is this?

- [x] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/integration/test_mqtt.py [hunk 1]
```diff
@@ -92,12 +94,64 @@ def callback(msg):
...
+
+
+def test_compas_data_as_message():
+
+    class DataTestMessage(Data):
+        def __init__(self, name=None, eclipse=None):
+            self.name = name
+            self.eclipse = eclipse
+
+        def __to_data__(self):
```
Comment (chenkasirer) [inline tests/integration/test_mqtt.py:116]: ah is this not a re-implementation of `__data__`?

**Review by chenkasirer — APPROVED** [review]
Generally LTGM, but I am a bit confused by some things. Mostly the (de)serialization of `Message` objects, just seems like a parallel system to the `compas.data`. But if it's nonsense feel free to go ahead :)

File: src/compas_eve/ghpython/background.py [hunk 2]
```diff
@@ -54,16 +54,27 @@ def do_something_long_and_complicated(worker):
...
+        and resource deallocation.
+    auto_set_done : bool, optional
+        If true, the worker state will be automatically set to ``Done`` after the function returns.
+        Defaults to ``True``.
+    args : tuple, optional
+        List or tuple of arguments for the invocation of the ``long_running_function``. Defaults to ``()``.
     """
 
-    def __init__(self, ghenv, long_running_function=None):
+    def __init__(self, ghenv, long_running_function=None, dispose_function=None, auto_set_done=True, args=()):
```
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_eve/ghpython/background.py:67]: ah but you already have kwargs going on..

File: src/compas_eve/core.py [hunk 3]
```diff
@@ -103,6 +113,26 @@ def __init__(self, name, message_type, **options):
...
+        However, it can deal with regular dictionaries as well as classes
+        implementing the COMPAS data framework.
+        """
+        try:
+            data = message.data
+        except (KeyError, AttributeError):
+            try:
+                data = dict(message)
+            except Exception:
+                data = message.__to_data__()
```
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/core.py:129]: yep, you're totally right, i fucked up the naming, it should be `__data__`

File: src/compas_eve/core.py [hunk 4]
```diff
@@ -103,6 +113,26 @@ def __init__(self, name, message_type, **options):
         self.message_type = message_type
         self.options = options
 
+    def _message_to_json(self, message):
+        """Convert a message to a JSON string.
```
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/core.py:117]: they are doing something similar to `compas.data` but not exactly, a message can be a simple dictionary, an instance of `Message`, or a subclass of `compas.data.Data` and needs to support compas 1.x and compas 2.x

File: src/compas_eve/core.py [hunk 5]
```diff
@@ -103,6 +113,26 @@ def __init__(self, name, message_type, **options):
...
+        Normally, this method expects sub-classes of ``Message`` as input.
+        However, it can deal with regular dictionaries as well as classes
+        implementing the COMPAS data framework.
+        """
+        try:
+            data = message.data
+        except (KeyError, AttributeError):
+            try:
+                data = dict(message)
+            except Exception:
```
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/core.py:128]: true, I was too lazy to think which one is the correct exception here :P

File: src/compas_eve/ghpython/background.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/ghpython/background.py:67]: not quite in this case, these `args` are the same as the `threading.Thread()` args, they are a list/tuple of arguments passed to the function later on

File: src/compas_eve/ghpython/background.py [hunk 6]
```diff
@@ -86,17 +97,32 @@ def start_work(self):
...
 
         target = _long_running_task_wrapper
         args = (self,)
         self.thread = threading.Thread(target=target, args=args)
+        self.thread.daemon = True
         self.thread.start()
 
+    def dispose(self):
+        """Invoked when the worker is being disposed."""
+        if self.dispose_function:
```
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/ghpython/background.py:123]: good point

File: tests/integration/test_mqtt.py (hunk 1 above)
Comment (gonzalocasas) [reply tests/integration/test_mqtt.py:116]: sorry, i was wrong again, it should be `__data__` indeed

File: src/compas_eve/core.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/core.py]: fixed at https://github.com/compas-dev/compas_eve/pull/12/commits/d2caf2c3e5511a2b0c5889074d51a1b6cbe0f3dd

File: src/compas_eve/ghpython/background.py [hunk 7]
```diff
@@ -86,17 +97,32 @@ def start_work(self):
...
-                worker.set_internal_state_to_done(result)
+                result = self.long_running_function(self, *self.args)
+
+                # If a long running function is not so long-running,
+                # it don't want to set done automatically
+                # e.g. a subscriber function will terminate immediately after
+                # setting up event handlers, but it's not DONE.
+                # It would be possible to busy-wait in those cases,
+                # but that's kind of silly, instead, we have this flag to control
+                # the state setting to be manual
```
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_eve/ghpython/background.py]: Sorry, I re-read the text and it was totally broken english. I rewrote it entirely now.

**Comment by gonzalocasas** [comment]
> Generally LTGM, but I am a bit confused by some things. Mostly the (de)serialization of `Message` objects, just seems like a parallel system to the `compas.data`. But if it's nonsense feel free to go ahead :)

Thanks! I think I fixed all the review findings. About the (de)serialization of `Message`, the idea is not to go entirely parallel to `compas.data`, but to support it without forcing it to be the case. So, a message can be an instance of `Message` without implementing `compas.data.Data` as well. Ideally, it's also possible to send/receive even a simple dictionary.

**Pull request merged**